### ISSUE TITLE
Extract language agnostic code

### ISSUE THREAD

**Issue by marcospassos** [felixfbecker/php-language-server#304]
I would be nice to split this awesome library into (at least) two libraries. I work at Croct and we have a proprietary language which we want to implement this protocol. For us, it makes sense to reuse the infrastructure classes, such as the protocol stubs, among other classes

**Comment by felixfbecker** [comment]
Is there a reason you want to implement it in PHP? The only reason I implemented this in PHP is because the best PHP parsers are written in PHP :)

**Comment by marcospassos** [comment]
Except for the parser, our entire codebase is written in PHP. Static analysis, evaluation and code instrumentation is also written in PHP.

My point is that a relevant part of this library could be reused by other languages to implement the language protocol using PHP, but not for instrumentating PHP.

**Comment by felixfbecker** [comment]
Yeah, was just curious. I'm totally open to moving all the general LSP stuff into its own package. I just created an empty repository, if you wann do a PR to that that extracts the LSP stuff: https://github.com/felixfbecker/php-language-server-protocol

**Comment by dantleech** [comment]
:wave: hey, I've separated the Protocol into it's own package, but can't make a PR because the "repository is empty", could you perhaps create an empty file in it or similar? cheers!

**Comment by dantleech** [comment]
Thanks :) I've made a PR: https://github.com/felixfbecker/php-language-server-protocol/pull/1 if it makes sense, I can make another one on this repo to integrate it.
